docs/proposals/configmap-rollout.md

```diff
@@ -0,0 +1,181 @@
+deployment "bar" updated
+```
+
+A new controller needs to watch for Deployments and ConfigMaps. Whenever it
+spots a Deployment with a `triggered-by` annotation and a mounted ConfigMap,
+it will create a copy of the ConfigMap and mount the copy in the Deployment.
+It should generally track Deployments with `triggered-by` annotations, 
+create new ConfigMap copies on updates of the original ConfigMap, and mount
+them back in the Deployment. Only Deployments should work for this controller
+for now but in the future the controller can be extended to facilitate config 
+management for DaemonSets, PetSets, and any other resource that may need it.
+
+Once the copied ConfigMap is mounted in the Deployment, the deployment
+controller needs to create the new ReplicaSet and set the owner reference
+for the copy to point to the ReplicaSet. In that way we ensure that copies
+of the original ConfigMap will be pruned by the garbage collector once their
+owner ReplicaSets are deleted. Certain copies of a ConfigMap will end up 
+being shared by more than one ReplicaSets.
+
+Additionally, the deployment controller should not create a new ReplicaSet
+for a Deployment "bar" that holds a `triggered-by` annotation to a ConfigMap
+"foo". Instead, it should wait until the new controller updates the Deployment
+template with the first copy of "foo". This ensures that the original ConfigMap
+copy will not land in a ReplicaSet, hence will not be subject to garbage
+collection.
+
+
+### Alternative solutions
+
+* Embed ConfigMap as a template in the Deployment.
+
+While it will work smoothly on ConfigMap changes for Deployments, this
+solution will need to be spread across the API for all resources that face
+the same problem. Sooner or later, higher-level resources will get bloated
+trying to manage stuff like ConfigMaps, Secrets, et al.
+
+* Others?
+
+## Changes
+
+### API
+
+No API changes will be needed for Deployments or ConfigMaps.
+
+### Controller
+
+A new controller will be needed for watching Deployments and Config Maps. Changes will also be needed in the deployment controller. [See above](#proposed-solution).
+
+### CLI
+
+We need a way to specify relationship between a resource that is updated (ConfigMap in our case) and a resource that will be triggered on those updates (Deployment). A generic `set trigger` will be added in `kubectl`. The relationship will be established with an annotation on the triggered resource (Deployment).
+```sh
+$ kubectl set trigger deployment/bar configmap/foo
+deployment "bar" updated
+$ kubectl get deployment/bar -o yaml | grep triggered-by
+deployment.kubernetes.io/triggered-by: configmap/foo
+```
+
+We also need a way to set volumes in pod templates without having to edit files by hand. A `kubectl set volume` command will be added to accommodate that need.
+```sh
+$ kubectl set volume deployment/bar -t configmap --configmapname foo
+deployment "bar" updated
+```
+
+## Issues
```

Comment:
Certainly, bundling the CM into the Deployment would suit your need well since you can atomically update and run a new deployment but I don't consider this solution optimal long-term because we would need to do the same thing for secrets, and I am worried about the usability of Deployments once we overload them with all configuration objects there are. Are you more worried about the garbage collection of old config maps or the atomicity that is missed if we don't bundle? In the latter case, I would expect manual deployments to be the best option eg. you always have a paused Deployment, run all the updates you want (new image, updated cm, ...), and start one-off rollouts (pausing and resuming is not a solution to manual deployments really).